From b1d2cb7ff02c55d25fa2c1028ad45e729a643a51 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <zhengxingda@iscas.ac.cn>
Date: Tue, 5 May 2026 17:29:31 +0800
Subject: [PATCH 1/2] pvr: allow STORAGE usage for WSI images

WSI images are expected to accept the same usage with OPTIMAL tiling
images, and this is checked in the CTS dEQP-VK.wsi.*.headless.swapchain.
create.image_usage test.

As storage to 2D linear images seems to be okay, check whether the image
is a WSI one by Mesa-private pNext, and allow STORAGE usage for those
that are created by Mesa WSI.

Signed-off-by: Icenowy Zheng <zhengxingda@iscas.ac.cn>
---
 src/imagination/vulkan/pvr_formats.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/src/imagination/vulkan/pvr_formats.c b/src/imagination/vulkan/pvr_formats.c
index 5d5a6d982d9..4291e6568b0 100644
--- a/src/imagination/vulkan/pvr_formats.c
+++ b/src/imagination/vulkan/pvr_formats.c
@@ -629,6 +629,16 @@ pvr_get_image_format_properties(struct pvr_physical_device *pdevice,
       goto err_unsupported_format;
    }
 
+   /* Allow STORAGE usage for WSI linear images, which is expected by the
+    * Vulkan specification (because WSI images should be as capable as optimal
+    * images). WSI images are expected to be all 2D.
+    */
+   if (vk_find_struct_const(info->pNext, WSI_IMAGE_CREATE_INFO_MESA) &&
+       tiling == VK_IMAGE_TILING_LINEAR &&
+       (pvr_format->bind & PVR_BIND_STORAGE_IMAGE)) {
+      tiling_features2 |= VK_FORMAT_FEATURE_2_STORAGE_IMAGE_BIT;
+   }
+
    /* If VK_IMAGE_CREATE_EXTENDED_USAGE_BIT is set, the driver can't decide if a
     * specific format isn't supported based on the usage.
     */
@@ -810,7 +820,8 @@ VkResult pvr_GetPhysicalDeviceImageFormatProperties2(
 
    /* Extract input structs */
    vk_foreach_struct_const (ext, pImageFormatInfo->pNext) {
-      switch (ext->sType) {
+      /* Casting to accept Mesa-private types w/o compiler warning */
+      switch ((unsigned)ext->sType) {
       case VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_EXTERNAL_IMAGE_FORMAT_INFO:
          external_info = (const void *)ext;
          break;
@@ -818,6 +829,7 @@ VkResult pvr_GetPhysicalDeviceImageFormatProperties2(
          break;
       case VK_STRUCTURE_TYPE_IMAGE_STENCIL_USAGE_CREATE_INFO:
       case VK_STRUCTURE_TYPE_PHYSICAL_DEVICE_IMAGE_DRM_FORMAT_MODIFIER_INFO_EXT:
+      case VK_STRUCTURE_TYPE_WSI_IMAGE_CREATE_INFO_MESA:
          /* Nothing to do here, it's handled in
           * pvr_get_image_format_properties)
           */
-- 
2.52.0

From fafad66af159db0c7deae29899d511061bae7b49 Mon Sep 17 00:00:00 2001
From: Icenowy Zheng <zhengxingda@iscas.ac.cn>
Date: Wed, 27 May 2026 20:50:44 +0800
Subject: [PATCH 2/2] pvr: advertise BXM-4-64 as conformant

CTS tests have passed on BXM-4-64 (both 36.52.104.182 and 36.56.104.183)
with the open source driver.

Advertise them as conformant.

Signed-off-by: Icenowy Zheng <zhengxingda@iscas.ac.cn>
---
 src/imagination/vulkan/pvr_physical_device.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/imagination/vulkan/pvr_physical_device.c b/src/imagination/vulkan/pvr_physical_device.c
index 51af21a832c..93e775a01a1 100644
--- a/src/imagination/vulkan/pvr_physical_device.c
+++ b/src/imagination/vulkan/pvr_physical_device.c
@@ -1056,7 +1056,9 @@ static bool pvr_device_is_conformant(const struct pvr_device_info *info)
 {
    const uint64_t bvnc = pvr_get_packed_bvnc(info);
    switch (bvnc) {
+   case PVR_BVNC_PACK(36, 52, 104, 182):
    case PVR_BVNC_PACK(36, 53, 104, 796):
+   case PVR_BVNC_PACK(36, 56, 104, 183):
       return true;
 
    default:
-- 
2.52.0

